### PR TITLE
Idempotent object manipulation modules.

### PR DESCRIPTION
**Pull request by mrichardson03**
Created new modules:
- panos_address_object
- panos_address_group_object
- panos_service_object
- panos_service_group_object
- panos_tag_object
- panos_object_facts

These replace the following modules:
- panos_object
- panos_sag
- panos_dag

### CODE REVIEW

**Comment by shinmog** [comment]
Some feedback in general, still going through everything:

* Please remove the trailing `_object` from the address group and service group modules (will need to update the example yaml files that reference these, too).
* In the example yaml files, please change the authentication variable names to `ip_address` / `username` / `password` like the other example files.
* In the facts module, I think you can simplify the params to one instead of two.  So, instead of `type` and `name`, you'd have `address_object`, `address_group`, etc, where the param names match the module names.
* You have device group considerations if the user gives you a Panorama IP address, but not vsys considerations if it's a firewall.  Please add a `vsys` param to all of these modules (default: `vsys1`) and set the device's vsys if it's a firewall.

**Comment by shinmog** [comment]
Oh, also changed "version_added" from 2.7 to 2.8 globally.

**Comment by shinmog** [comment]
Another thing I've seen in a few modules:  don't name variables `type`.  This is a reserved keyword in python, and defining a local variable masks that builtin.

**Comment by shinmog** [comment]
Another pull-wide comment - modules should all have a `commit` bool that defaults to `true`.

**Comment by mrichardson03** [comment]
> Another pull-wide comment - modules should all have a `commit` bool that defaults to `true`.

I disagree.  The `panos_commit` module should be leveraged to do commits.

**Comment by mrichardson03** [comment]
> Some feedback in general, still going through everything:
> 
> * Please remove the trailing `_object` from the address group and service group modules (will need to update the example yaml files that reference these, too).
> * In the example yaml files, please change the authentication variable names to `ip_address` / `username` / `password` like the other example files.
> * In the facts module, I think you can simplify the params to one instead of two.  So, instead of `type` and `name`, you'd have `address_object`, `address_group`, etc, where the param names match the module names.
> * You have device group considerations if the user gives you a Panorama IP address, but not vsys considerations if it's a firewall.  Please add a `vsys` param to all of these modules (default: `vsys1`) and set the device's vsys if it's a firewall.

I can do all of this except vsys support.  I do not have access to a vsys-capable firewall for testing.

File: library/panos_address_object.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
+#!/usr/bin/env python
+
+#  Copyright 2017 Palo Alto Networks, Inc
```
[the remark this answers is not in the document]
Comment (mrichardson03) [reply library/panos_address_object.py]: The `type` arg in `panos_address_object` is consistent with the args used in [pandevice](https://pandevice.readthedocs.io/en/latest/module-objects.html#pandevice.objects.AddressObject).  I will change it in `panos_address_group`.

I've got test playbooks that create shared objects with the code as-is.

**Comment by shinmog** [comment]
Regarding the `type` param:  I'm going to do my best to make sure that there are no more params in pandevice from here on out named `type`.  But if the param is named `type`, we shouldn't define a variable of that name in the module as it hides the `type` builtin, so we're forced to name it something other than what the pandevice param is.

Regarding the "shared" scope:  implicitly creating an object in "shared" by not specifying a device group in the playbook will work as your code is right now, but if the user explicitly asks for "shared" as the device group it will not work.

**Comment by shinmog** [comment]
Ansible galaxy linter; maybe don't worry about this, as probably every file will need to be touched anyways..

**Comment by shinmog** [comment]
Perfect, thanks for all the hard work! 👍

**Pull request merged**